test/lock_tables.py

```diff
@@ -0,0 +1,150 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+import logging
+import os
+import unittest
+
+import MySQLdb
+
+import environment
+import tablet
+import utils
+from utils import TestError
+from urlparse import urlparse
+import grpc
+from vtproto.tabletmanagerdata_pb2 import LockTablesRequest, UnlockTablesRequest
+from vtproto.tabletmanagerservice_pb2_grpc import TabletManagerStub
+
+use_mysqlctld = False
+
+tablet_master = None
+tablet_replica1 = None
+
+
+def setUpModule():
+    global tablet_master, tablet_replica1
+
+    tablet_master = tablet.Tablet()
+    tablet_replica1 = tablet.Tablet()
+    try:
+        environment.topo_server().setup()
+
+        # start mysql instance external to the test
+        setup_procs = [
+            tablet_master.init_mysql(),
+            tablet_replica1.init_mysql(),
+        ]
+        if use_mysqlctld:
+            tablet_master.wait_for_mysqlctl_socket()
+            tablet_replica1.wait_for_mysqlctl_socket()
+        else:
+            utils.wait_procs(setup_procs)
+    except:
+        tearDownModule()
+        raise
+
+
+def tearDownModule():
+    utils.required_teardown()
+    if utils.options.skip_teardown:
+        return
+
+    teardown_procs = [
+        tablet_master.teardown_mysql(),
+        tablet_replica1.teardown_mysql(),
+    ]
+    utils.wait_procs(teardown_procs, raise_on_error=False)
+
+    environment.topo_server().teardown()
+    utils.kill_sub_processes()
+    utils.remove_tmp_files()
+
+    tablet_master.remove_tree()
+    tablet_replica1.remove_tree()
+
+
+class TestLockTables(unittest.TestCase):
+
+    def setUp(self):
+        for t in tablet_master, tablet_replica1:
+            t.create_db('vt_test_keyspace')
+
+        tablet_master.init_tablet('replica', 'test_keyspace', '0', start=True)
+        tablet_replica1.init_tablet('replica', 'test_keyspace', '0', start=True)
+        utils.run_vtctl(['InitShardMaster', '-force', 'test_keyspace/0',
+                         tablet_master.tablet_alias])
+
+    def tearDown(self):
+        for t in tablet_master, tablet_replica1:
+            t.kill_vttablet()
+
+        tablet.Tablet.check_vttablet_count()
+        environment.topo_server().wipe()
+        for t in [tablet_master, tablet_replica1]:
+            t.reset_replication()
+            t.set_semi_sync_enabled(master=False, slave=False)
+            t.clean_dbs()
+
+    _create_vt_insert_test = '''create table vt_insert_test (
+  id bigint auto_increment,
+  msg varchar(64),
+  primary key (id)
+  ) Engine=InnoDB'''
+
+    def test_lock_and_unlock(self):
```

Comment:
should we also have a test for auto-unlock on timeout? Maybe if you expose that as a flag then set it to something really low like 1 second when you start the tablet and then you sleep for 2 seconds and the tables should be unlocked?

Reply:
https://github.com/vitessio/vitess/pull/4310/commits/0a242beb176b1711323316ed187bdd5cb55973b0